public/modules/custom/grants_handler/src/ApplicationHandler.php

```diff
@@ -557,6 +556,75 @@ public static function createApplicationNumber(WebformSubmission $submission): s
     return 'GRANTS-' . $appParam . '-' . $typeCode . '-' . sprintf('%08d', $serial);
   }
 
+  /**
+   * Generate next available application number for the submission.
+   *
+   * @param \Drupal\webform\Entity\WebformSubmission $submission
+   *   Webform data.
+   *
+   * @return string
+   *   Generated number.
+   */
+  public static function getAvailableApplicationNumber(WebformSubmission &$submission): string {
+    $appParam = self::getAppEnv();
+    $serial = $submission->serial();
+    $webform_id = $submission->getWebform()->id();
+    $lastSerialKey = $webform_id . '_' . $appParam;
+    $kvService = \Drupal::service('keyvalue.database');
+    $kvStorage = $kvService->get('application_numbers');
+    $savedSerial = $kvStorage->get($lastSerialKey);
+
+    if (!empty($submission->getData())) {
+      return self::createApplicationNumber($submission);
+    }
+
+    if ($savedSerial && $savedSerial > $serial) {
+      $serial = $savedSerial;
+    }
+
+    $atvService = \Drupal::service('helfi_atv.atv_service');
+
+    $applicationType = $submission->getWebform()
+      ->getThirdPartySetting('grants_metadata', 'applicationType');
+
+    $typeCode = self::getApplicationTypes()[$applicationType]['code'] ?? '';
+
+    $check = FALSE;
+    while ($check === FALSE) {
+      $applicationNumber = self::getApplicationNumberInEnvFormat($appParam, $typeCode, $serial);
+      $applNumberIsAvailable = $atvService->checkDocumentExistsByTransactionId($applicationNumber);
+      if ($applNumberIsAvailable) {
+        // Check that there is no local submission with given serial.
+        $query = \Drupal::entityQuery('webform_submission')
+          ->condition('webform_id', $webform_id)
+          ->condition('serial', $serial);
+        $results = $query->execute();
+
+        if (empty($results)) {
+          $submission->set('serial', $serial);
+          $kvStorage->set($lastSerialKey, $serial);
+          return $applicationNumber;
+        }
+      }
+      // No luck, let's check another one.
+      $serial++;
+    }
+
```

Comment:
![image](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/892920/2ff83bfc-0d59-444e-8cc3-e7f577c45d34)

Pitäsköhän ton palauttaa aina jotain kuitenki?